### DISCUSSION TITLE
When do I need `placeholder`?

### DISCUSSION THREAD

**Discussion by Berhtulf** [pointfreeco/swift-issue-reporting#38] [Q&A]
This might be a silly question, but when exactly do I have to provide a `placeholder`? I am looking at the isowords examples and there are all sorts of calls in Clients... sync/async, throwing, returning Void, returning values etc.

But I can't figure out when do I have to add the placeholder. Only time I get some kind of feedback is the fatalError during testing, if a function is called without specifying test dependency version.

Is it possible to provide some guidance in the documentation, `XCTUnimplemented` initialization or some compile time checks?

**Comment by stephencelis** [comment]
Generally, `unimplemented` will try to fail and not use a placeholder. For example, all throwing functions that are overridden using `unimplemented` will simply throw. For non-throwing versions, a placeholder value can be set to be returned, but can be omitted, in which case the library will attempt to generate a placeholder, or fail with a `fatalError`.

We mainly support automatic placeholders for `Void` values and for empty collections: https://github.com/pointfreeco/xctest-dynamic-overlay/blob/main/Sources/XCTestDynamicOverlay/Unimplemented.swift#L547-L553

@iampatbrown Opened a PR to add more defaults, though, which we should maybe revisit: https://github.com/pointfreeco/xctest-dynamic-overlay/pull/25
